content/docs/user-guide/project-structure/pipelines-files.md

```diff
@@ -0,0 +1,416 @@
+    vars:
+      - params.json:build
+      - model:
+          filename: 'model-us.hdf5'
+    cmd: python train.py ${build.epochs} --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+⚠️ Known limitations of local `vars`:
+
+- [`wdir`](/doc/user-guide/dvc-files/dvc.yaml#accepted-fields) cannot use values
+  from local `vars`, as DVC uses the working directory first (to load any values
+  from params files listed in `vars`).
+- `foreach` is also incompatible with local `vars` at the moment.
+
+The substitution expression supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the expression in `dvc.yaml`, for example for the shell to interpret a
+> command substitution, escape it with a backslash, e.g. `\${...`.
+
+## Generating multiple stages
+
+You can define more than one stage in a single `dvc.yaml` entry with the
+following syntax. A `foreach` element accepts a list or dictionary with values
+to iterate on, while `do` contains the regular stage fields (`cmd`, `outs`,
+etc.). Here's a simple example:
+
+```yaml
+stages:
+  cleanups: # Multi-stage
+    foreach: # List of simple values
+      - raw1
+      - labels1
+      - raw2 labels2
+    do:
+      cmd: clean.py "${item}"
+      out:
+        - ${item}.cln
+```
+
+Upon `dvc repro`, each item in the list is expanded into its own stage by
+substituting its value in expression `${item}`. The item's value is appended to
+each stage name after a `@`. The final generated stages are saved to `dvc.lock`:
+
+```yaml
+cleanups@raw2 labels2:
+  cmd: echo "raw2 labels2"
+cleanups@raw1:
+  cmd: echo "raw1"
+cleanups@labels1:
+  cmd: echo "labels1"
+```
+
+For lists containing complex values (e.g. dictionaries), the substitution
+expression can use the `${item.key}` form. Stage names will be appended with a
+zero-based index. For example:
+
+```yaml
+stages:
+  train:
+    foreach:
+      - epochs: 3
+        thresh: 10
+      - epochs: 10
+        thresh: 15
+    do:
+      cmd: python train.py ${item.epochs} ${item.thresh}
+```
+
+```yaml
+# dvc.lock
+stages:
+  train@0:
+    cmd: python train.py 3 10
+  train@1:
+    cmd: python train.py 10 15
+```
+
+DVC can also iterate on a dictionary given directly to `foreach`, resulting in
+two substitution expressions being available: `${key}` and `${item}`. The former
+is used for the stage names:
+
+```yaml
+stages:
+  build:
+    foreach:
+      uk:
+        epochs: 3
+        thresh: 10
+      us:
+        epochs: 10
+        thresh: 15
+    do:
+      cmd: python train.py '${key}' ${item.epochs} ${item.thresh}
+      outs:
+        - model-${key}.hdfs
+```
+
+```yaml
+# dvc.lock
+stages:
+  build@uk:
+    cmd: python train.py 'uk' 3 10
+    outs:
+      - model-uk.hdfs
+  build@us: ...
+```
+
+Importantly, dictionaries [from parameters](#templating) files can be used in
+`foreach` multi-stages as well:
+
```

Comment:
Could an example `params.yaml` be included here? I'm not totally clear how this works. I'm assuming `params.yaml` would look like:

```yaml
myobject:
  us:
    prop1: 3
    prop2: 10
  uk:
    prop1: 10
    prop2: 15
```

Reply:
I fear it'll be too many code blocks... For now I changed the link to go to https://dvc.org/doc/command-reference/params instead. At some point (hopefully soon) those links will go to a new Concepts section with the explanation and better examples of each thing (in this case parameters files) — see #550. WDYT?

Reply:
Sounds good. Once a Concepts section exists, it would be great to have the templating example use the same params that are shown in the example `params.yaml`.